File: Client/Services/CartProductsGroupService.cs
namespace GroceryListHelper.Client.Services;

public sealed class CartProductsGroupService : ICartProductsService
{
    private readonly HttpClient client;
    private readonly string uri;

    public CartProductsGroupService(IHttpClientFactory clientFactory, NavigationManager navigation)
    {
        client = clientFactory.CreateClient("ProtectedClient");
        string groupId = navigation.Uri[navigation.Uri.LastIndexOf('/')..];
        uri = $"api/cartgroupproducts{groupId}";
    }

    public async Task<List<CartProductUIModel>> GetCartProducts()
    {
        return await client.GetFromJsonAsync<List<CartProductUIModel>>(uri) ?? new List<CartProductUIModel>();
    }

    public async Task SaveCartProduct(CartProduct product)
    {
        HttpResponseMessage response = await client.PostAsJsonAsync(uri, product);
        response.EnsureSuccessStatusCode();
    }

    public async Task DeleteCartProduct(string productName)
    {
        HttpResponseMessage response = await client.DeleteAsync(uri + $"/{productName}");
        response.EnsureSuccessStatusCode();
    }

    public async Task DeleteAllCartProducts()
    {
        await client.DeleteAsync(uri);
    }

    public async Task UpdateCartProduct(CartProductCollectable cartProduct)
    {
        await client.PutAsJsonAsync(uri, cartProduct);
    }

    public async Task SortCartProducts(ListSortDirection sortDirection)
    {
        await client.PatchAsync(uri + "/sort/" + (int)sortDirection, null);
    }
}

File: Client/src/services/CartProductsGroupService.ts
import { getConnectionId, joinGroup } from "../helpers/cartHubClient";
import { get, send } from "../helpers/httpHelpers";
import type { CartProduct } from "../types/CartProduct";
import type { ICartProductsService } from "../types/ICartProductsService";
import type { SortDirection } from "../types/SortState";

export class CartProductsGroupService implements ICartProductsService {
    readonly url = "api/cartgroupproducts/";

    getCartProducts = async () => {
        await joinGroup(this.getGroupId());
        const productsResponse = await get(this.url + this.getGroupId(), "GET", async (r) => await r.json() as CartProduct[]);
        return productsResponse;
    }

    createCartProduct = async (product: CartProduct) => {
        await joinGroup(this.getGroupId());
        const response = await send(this.url + this.getGroupId(), "POST", product, { 'ConnectionId': getConnectionId() });
        return response;
    }

    updateCartProduct = async (product: CartProduct) => {
        await joinGroup(this.getGroupId());
        const response = await send(this.url + this.getGroupId(), "PUT", product, { 'ConnectionId': getConnectionId() });
        return response;
    }

    sortCartProducts = async (sortDirection: SortDirection) => {
        await joinGroup(this.getGroupId());
        const response = await send(this.url + this.getGroupId() + "/sort/" + (sortDirection === "Ascending" ? "0" : "1"), "PATCH", undefined, { 'ConnectionId': getConnectionId() });
        return response;
    }

    deleteAllCartProducts = async () => {
        await joinGroup(this.getGroupId());
        const response = await send(this.url + this.getGroupId(), "DELETE", undefined, { 'ConnectionId': getConnectionId() });
        return response;
    }

    deleteCartProduct = async (name: string) => {
        await joinGroup(this.getGroupId());
        const response = await send(this.url + this.getGroupId() + "/" + name, "DELETE", undefined, { 'ConnectionId': getConnectionId() });
        return response;
    }

    getGroupId = () => {
        const lastSlashIndex = window.location.href.lastIndexOf('/');
        const result = window.location.href.substring(lastSlashIndex + 1);
        return result;
    }
}